### ISSUE TITLE
Deregister core connections and mutations

### ISSUE THREAD

**Issue by stubbies** [wp-graphql/wp-graphql#1733]
It would be great if we could deregister some core connections, in our case we are using only a small set of queries for posts, custom posts, menus. When generating types on the js-client we get a lot of unused types like enum `TimezoneEnum`.

**Comment by CodeProKid** [comment]
Hey @stubbies you can in fact deregister the fields & connections that come with the plugin by default. Check out the `deregister_graphql_field` function: https://www.wpgraphql.com/functions/deregister_graphql_field/.

**Comment by ramyareye** [comment]
I think I need same thing, and it's not deregister field but deregister model(object)!
I mean as @stubbies said remove other models from hole schema. any idea? something like `deregister_graphql_model` or `deregister_graphql_object`

**Comment by justlevine** [comment]
While reviewing #874 , I couldn't find a deregister any GraphQL _types_ ( `deregister_graphql_field()` only unhooks the field from the parent type ), a prerequisite for any type-based deregistration (e.g objects or mutations) .

This isn't an easy feat, because in order to _fully_ remove a type from the schema you also need to:
1. Remove any fields that resolve to that type.
2. Remove any connections that resolve to that type.
3. Prevent any interfaces/unions from resolving to that that type.
4. Clean up any types that no longer have any fields or interfaces/unions that no longer have any types to resolve to (repeating 1-4 for each).

Before dealing with the recursive logic, here's what I gleaned so far:

### Deregistering mutations
This is more of a disclaimer. `graphql_deregister_field( 'RootMutation', $mutation )` seems to deregister `${mutation}Input` and `${mutationPayload}` but I cant figure out how/why. Understanding this might provide a alternatives/solutions to what follows:

### Deregistering the type itself
At the top of `TypeRegistry->register_type()`, we can bail early if `graphql_should_register_type` filter is false.
```php
public function register_type( string $type_name, $config ) {
  /**
   * Filter whether the GraphQL type should be registered to the schema.
   *
   * @param bool $should_register Whether the GraphQL type should be registered to the schema.
   * @param string $type_name The name of the the GraphQL type to register.
   * @param mixed|array|Type $config The config for the type.
   */
  if ( ! apply_filters( 'graphql_should_register_type', true, $type_name, $config ) ) {
    return;
  }
  // rest of function
}
```

It's simple, but without cleaning up any associations, you get those impossible-to-debug `GraphQL\\Type\\Schema::resolveType(): Return value must be of type GraphQL\\Type\\Definition\\Type, null returned` errors.

### Deregistering connections
Connections dont inherit `TypeRegistry->register_type()`, so without a big refactor, we need to add a similar filter to the one above:

```php
public function register_connection( array $config ) {
  if ( ! apply_filters( 'graphql_should_register_connection', true, $config['fromType'], $config['toType'], $config['fromFieldName'], $config ) ) {
    return;
  }
  $connection = new WPConnectionType( $config, $this );
  $connection->register_connection();
}
```
Since `WPConnectionType->register_connection()` is responsible for the connection inputs and field, we dont need to to worry about cleaning up those generated types. We likely do need a way to clean up 3rd party plugins trying to register fields _to_ those generated types.

### Union/Interface type resolvers
I'm stumped, as each interface/union has their own `resolveType` callable we need to clean up. We can overwrite the entire callable with `graphql_{interface|union}_resolve_type`, but I cant see an obvious way to short-circuit the the callable _before_ it tries to resolve to the deregistered type.

### Access functions
Not really much we can do until we figure out the above, but I think we're looking at something like the following:
```php
deregister_graphql_type( string $type_name ) : void {
  // Handle the connections.
 add_filter(
   'graphql_should_register_connection',
   function ( bool $should_register, string $from_type, string $to_type, string $from_type_name ) {
     return ucfirst( $from_type ) === $type_name || ucfirst( $to_type ) === $type_name ? false : $should_register;
   },
   10,
   4
  );
  // handle the interface/union type resolvers somehow

  // handle the types
  add_filter(
    'graphql_should_register_type',
    function ( bool $should_register, string $name ) use ( $type_name ) {
      // do we clean up the fields here?
      return ucfirst( $name ) === $type_name ? false : $should_register;
    },
    10,
    2
  );
  // or do we clean up the fields here?
}
```

We still need a way to loop through all the type in the schema and deregister any fields that use those types. Seemingly along the lines of 
```php
// loop through all the still-registered types, grab their configs, and then check each field type against ours.
foreach ( $config['fields'] as $field_name => $field_config ) {
  if ( empty( $field_config['type'] ) ) {
    break;
  }
  // Gets the field type name without any modifiers (e.g. "non-null" ).
  $field_type = Utils::get_field_type_name_from_definition( $field_config['type'] ); // 
  if ( ucfirst( $field_type ) === $type_name ) {
    deregister_graphql_field( $name, $field_name ); // Does the action work correctly here?
  }
}
...
// Get the type name without any modifiers (e.g. "non-null").
public static function get_field_type_name_from_definition( $type ) {
  if ( ! is_array( $type ) ) {
    return $type;
  }

  return self::get_field_type_name_from_definition( array_values( $type )[0] );
}
```
Keep in mind that the WPGraphQL lifecycle isnt very well documented. Adding actions in filters and vice versa is likely to cause headaches. This is true for when we try to clean any references to the type (`deregister_graphql_field()` uses an action, without it we need to performantly get and loop through every single type over and over), and double true when we need to get recursive.

## Tl;dr
Afaik, deregistering a GraphQL type in and of itself isnt so hard, however I cant figure out a clear way to purge all the references (fields, type resolvers) to the deregistered type.
Maybe there is something simple I'm overlooking (a la deregistering mutation fields that automatically also gets rid of `${mutation}Payload` and `${mutation}Input`, but if not this will probably take some major refactoring.

Thoughts?
